OscLib/OscCalcNuFast.cxx

```diff
@@ -0,0 +1,295 @@
+  VT _OscCalcNuFast<T>::_P(int from, int to, KVT E) {
+    // First, make sure the PDG codes have the same sign.
+    assert((from > 0 && to > 0) || (from < 0 && to < 0) );
+    // Nice trick from analytic: If we need antineutrinos, just flip flavor signs and send E -> -E. The NuFast algorithm already uses -E means antineutrinos.
+    if ( from < 0 ) { return P(-from,-to,-E); }
+    
+    // The probability cache was invalidated if the user updated any parameters.
+    if ( this->fIsDirty ) {
+      ClearProbCaches();
+      this->fIsDirty = false;
+    }
+    
+    // Try to find the requested energy in the probabilities cache. E is possibly negative (for antineutrinos).
+    typename analytic::ProbCache<KVT,VT>::iterator it = analytic::ProbCache<KVT,VT>::find(E);
+    if ( it != analytic::ProbCache<KVT,VT>::end() ) {
+      return it->second.P(from,to);
+    }
+    
+    // Need to do energy-dependent computation.
+    return this->RecomputeProbabilityMatrix<VT,KVT>(from,to,E);
+  }
+  
+  template<typename T> T _OscCalcNuFast<T>::P(int from, int to, double E) {
+    //std::cout << "NuFast: Calling double overload." << std::endl;
+    return _P<T>(from,to,E);
+  }
+  
+  template<typename T> Eigen::ArrayX<T> _OscCalcNuFast<T>::P(int from, int to, const Eigen::ArrayXd& E)
+  {
+    //std::cout << "NuFast: Calling Eigen::ArrayXd overload." << std::endl;
+    return _P<Eigen::ArrayX<T>>(from,to,E);
+  }
+  
+  template<typename T> Eigen::ArrayX<T> _OscCalcNuFast<T>::P(int from, int to, const std::vector<double>& E)
+  {
+    //std::cout << "NuFast: Calling std::vector<double> overload (size = " << E.size() << ")." << std::endl;
+    return P(from,to,Eigen::Map<const Eigen::ArrayXd>(E.data(),E.size()));
+  }
+  
+  template<typename T> template<class VT, class KVT>
+  VT _OscCalcNuFast<T>::RecomputeProbabilityMatrix(int from, int to, KVT E) {
+    // Now we need to interface internally used NuFast variables with the OscCalc members.
+    #define s12sq this->fPre.s12sq
+    #define s13sq this->fPre.s13sq
+    #define s23sq this->fPre.s23sq
+    #define sind this->fPre.sind
+    #define cosd this->fPre.cosd
+    #define c12sq this->fPre.c12sq
+    #define c13sq this->fPre.c13sq
+    #define c23sq this->fPre.c23sq
+    
+    #define delta this->fdCP
+    #define Dmsq21 this->fDmsq21
+    #define Dmsq32 this->fDmsq32
+    #define L this->fL
+    #define rho this->fRho
+    #define Ye this->fYe
+    #define N_Newton this->fNNewton
+    #define probs_returned this->fCurrProbs
+    
+    // --------------------------------------------------------------------- //
+    // First calculate useful simple functions of the oscillation parameters //
+    // --------------------------------------------------------------------- //
+    
+    // NOvA doesn't use dmsq31 by convention, but we can compute it if we know dmsq32 and dmsq21.
+    const T
+    Dmsq31 = Dmsq32 + Dmsq21, // dmsq31 = msq3 - msq1 = (msq3-msq2) + (msq2-msq1) = dmsq32 + dmsq21.
```

Comment:
Might be worth a comment here that this works for both mass orderings because in IO `Dmsq32` and `Dmsq21` have opposite sign